pytorch_lightning/loggers/comet.py

```diff
@@ -225,18 +219,35 @@ def save_dir(self) -> Optional[str]:
         return self._save_dir
 
     @property
-    def name(self) -> str:
-        return str(self.experiment.project_name)
+    def name(self) -> Optional[str]:
+        # don't create an experiment if we don't have one
+        return self._experiment.project_name if self._experiment else self._project_name
```

Comment:
The tests fail because of this here. If the experiment does not exist and the project name is also None, it will return None, and the trainer tries to os.path.join(..., None), that does not work. I would keep it as before. The same applies to the version.

Reply:
The previous code was problematic because it would create an experiment object if None exists. In addition, if project_name is None and no project_name is configured by the user, `self._experiment.project_name` can also returns None.
Wandb logger seems to also return None in that case https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loggers/wandb.py#L140, is there a difference between the Wandb logger and the Comet logger that would make returning None for the Logger name problematic?

Reply:
I guess the wandb logger in the tests is called to create the experiment before accessing the name. You are right there is no difference in the definition there. None is problematic when the Trainer is trying to concat the path to the logger directory, so it is doing os.path.join(root, name, version, ...) and if some of these are None it throws. 

Why do you think it is problematic that the call to e.g. name or version creates an experiment if it does not exist? The fact that the constructor does not immediately create the experiment means it needs to be triggered as soon as the user tries to interact with the logger.

Reply:
I can think of at least 2 scenarios when implicit experiment creation would be problematic:

- In DDP mode, if one process which is not the rank zero one tries to access `name` or `version`, it will create an Experiment that will not receives metrics or parameters. This will slow down the process and add noise as the Experiment (almost empty) summary will be displayed at the end of the training.
- In DDP mode, if one experiment is created before the actual start of the training, we will have to recreate another experiment in the DDP processes anyway so that will slow down the training and generates extra output too. This particular scenario would still happens if the user tries to access `logger.experiment` before calling `Trainer.fit` but I think we should avoid this if we can. I've move the handling of experiment_name to delay the creation of the actual Experiment until we really need it.

As I said before, for the logger name, even if the experiment exists, there is still a chance that the `project_name` is None. If it is absolutely required for the Logger to returns a non-null string, I will have to think about a solution. Maybe returning a default string.

Reply:
on rank > 0 the experiments don't get called, see the decorator "rank_zero_experiment". 

> As I said before, for the logger name, even if the experiment exists, there is still a chance that the project_name is None. If it is absolutely required for the Logger to returns a non-null string, I will have to think about a solution. Maybe returning a default string.

Do you think it would be too strict to make the project name mandatory?

Reply:
I think returning a default name would be reasonable too. Some other loggers do that.

Reply:
I've update the code to returns a default name in case no project_name is set. I've also added a couple of tests, I'm not sure if current build failures are linked to my changes or not, I have the impression they are not, could you confirm it?

Reply:
I don't think so, try to merge master, then they will be triggered again.

Reply:
I've updated my PR with your review comments and merged master. I don't think the current failures are linked to my changes.